src/Lykke.Service.ExchangeConnector/Exchanges/Concrete/Shared/OrderBooksHarvesterBase.cs

```diff
@@ -127,19 +143,121 @@ public void Stop()
             }
-            var orderBooks = from si in OrderBookSnapshot
-                             group si by si.Symbol into g
-                             let asks = g.Where(i => !i.IsBuy).Select(i => new VolumePrice(i.Price, i.Size)).ToArray()
-                             let bids = g.Where(i => i.IsBuy).Select(i => new VolumePrice(i.Price, i.Size)).ToArray()
-                             let assetPair = BitMexModelConverter.ConvertSymbolFromBitMexToLykke(g.Key, CurrencyMappingProvider).Name
-                             select new OrderBook(ExchangeName, assetPair, asks, bids, DateTime.UtcNow);
+            var orderBooks = OrderBookSnapshots.Values
+                .Select(obs => new OrderBook(
+                    ExchangeName,
+                    ConvertSymbolFromExchangeToLykke(obs.AssetPair).Name,
+                    obs.Asks.Values.Select(i => new VolumePrice(i.Price, i.Size)).ToArray(),
+                    obs.Bids.Values.Select(i => new VolumePrice(i.Price, i.Size)).ToArray(),
+                    DateTime.UtcNow));
             _publishedToRabbit++;
+            
             foreach (var orderBook in orderBooks)
             {
                 await _newOrderBookHandler(orderBook);
             }
         }
 
+        protected abstract Task MessageLoopImpl();
+
+        protected async Task<OrderBookSnapshot> GetOrderBookSnapshot(string pair)
+        {
+            if (!OrderBookSnapshots.TryGetValue(pair, out var orderBook))
+            {
+                var message = "Trying to retrieve a non-existing pair order book snapshot " +
+                              $"for exchange {ExchangeName} and pair {pair}";
+                await Log.WriteInfoAsync(nameof(MessageLoopImpl), nameof(MessageLoopImpl), message);
+                throw new OrderBookInconsistencyException(message);
+            }
+
+            return orderBook;
+        }
+
+        protected async Task<ConcurrentDictionary<string, OrderBookItem>> GetOrdersList(string pair, bool isBuy)
+        {
+            var orderBookSnapshot = await GetOrderBookSnapshot(pair);
+            return isBuy
+                ? orderBookSnapshot.Bids
+                : orderBookSnapshot.Asks;
+        }
+
+        protected async Task HandleOrdebookSnapshotAsync(string pair, DateTime timeStamp, 
+            IEnumerable<OrderBookItem> orders)
+        {
+            var orderBookSnapshot = new OrderBookSnapshot(ExchangeName, pair, timeStamp);
+            AddOrUpdateOrders(orderBookSnapshot, orders);
+
+            await OrderBookSnapshotsRepository.SaveAsync(orderBookSnapshot);
+
+            await PublishOrderBookSnapshotAsync();
+        }
+
+        protected async Task HandleOrdersEventsAsync(string pair, 
+            OrderBookEventType orderEventType,
+            ICollection<OrderBookItem> orders)
+        {
+            var orderBookSnapshot = await GetOrderBookSnapshot(pair);
+
+            await OrderBookEventsRepository.SaveAsync(new OrderBookEvent
```

Comment:
Maybe it should be configurable. Not sure that all exchanges must save order books to a DB